File: api-java/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/catalog/usecase/GetValorantWeaponUseCase.java
package fr.huiitre.tools.modules.riot.valorant.application.catalog.usecase;

import java.util.Optional;

import org.springframework.stereotype.Service;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantWeaponRepository;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.view.ValorantWeaponView;

@Service
public class GetValorantWeaponUseCase implements SecuredUseCase {

    private final ValorantWeaponRepository weaponRepository;

    public GetValorantWeaponUseCase(ValorantWeaponRepository weaponRepository) {
        this.weaponRepository = weaponRepository;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.READ_ONLY;
    }

    public ValorantWeaponView execute(Long id) {
        return weaponRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Weapon not found: " + id));
    }
}

File: api/Modules/Riot/Valorant/Application/Catalog/Usecases/GetValorantWeaponUseCase.cs
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Views;

namespace Tools.Api.Modules.Riot.Valorant.Application.Catalog.Usecases;

public sealed class GetValorantWeaponUseCase(
    UseCaseAuthorizer authorizer,
    IValorantWeaponRepository weaponRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.ReadOnly;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    public async Task<ValorantWeaponView> Execute(long id)
    {
        var weapon = await weaponRepository.FindById(id);

        return weapon ?? throw AppException.NotFound(
            "VALORANT_WEAPON_NOT_FOUND",
            $"L'arme {id} est introuvable.");
    }
}
